python/ray/serve/_private/replica_scheduler/pow_2_scheduler.py

```diff
@@ -480,56 +538,111 @@ async def select_from_candidate_replicas(
-        for c in candidates:
+        get_queue_len_tasks = []
+        for r in replicas:
             t = self._loop.create_task(
-                c.get_queue_state(deadline_s=queue_len_response_deadline_s)
+                r.get_queue_len(deadline_s=queue_len_response_deadline_s)
             )
-            t.replica_id = c.replica_id
-            get_queue_state_tasks.append(t)
+            t.replica = r
+            get_queue_len_tasks.append(t)
 
         done, pending = await asyncio.wait(
-            get_queue_state_tasks,
+            get_queue_len_tasks,
             timeout=queue_len_response_deadline_s,
             return_when=asyncio.ALL_COMPLETED,
         )
         for t in pending:
+            result.append((t.replica, None))
             t.cancel()
             logger.warning(
-                f"Failed to get queue length from replica {t.replica_id} "
+                f"Failed to get queue length from replica {t.replica.replica_id} "
                 f"within {queue_len_response_deadline_s}s. If this happens repeatedly "
                 "it's likely caused by high network latency in the cluster. You can "
                 "configure the deadline using the "
                 "`RAY_SERVE_QUEUE_LENGTH_RESPONSE_DEADLINE_S` environment variable."
             )
 
-        chosen_replica_id = None
-        lowest_queue_len = math.inf
         for t in done:
             if t.exception() is not None:
+                result.append((t.replica, None))
                 msg = (
                     "Failed to fetch queue length for "
-                    f"replica {t.replica_id}: '{t.exception()}'"
+                    f"replica {t.replica.replica_id}: '{t.exception()}'"
                 )
                 # If we get a RayActorError, it means the replica actor has died. This
                 # is not recoverable (the controller will start a new replica in its
                 # place), so we should no longer consider it for requests.
                 if isinstance(t.exception(), RayActorError):
-                    self._replicas.pop(t.replica_id, None)
-                    self._replica_id_set.discard(t.replica_id)
+                    self._replicas.pop(t.replica.replica_id, None)
+                    self._replica_id_set.discard(t.replica.replica_id)
                     for id_set in self._colocated_replica_ids.values():
-                        id_set.discard(t.replica_id)
+                        id_set.discard(t.replica.replica_id)
                     msg += " This replica will no longer be considered for requests."
 
                 logger.warning(msg)
             else:
-                queue_len, accepted = t.result()
-                if accepted and queue_len < lowest_queue_len:
-                    chosen_replica_id = t.replica_id
+                queue_len = t.result()
+                result.append((t.replica, queue_len))
+                self._replica_queue_len_cache.update(r.replica_id, queue_len)
```

Comment:
[Nit] We calculate the timestamp for all the responses upon update rather than when we actually received the response. This means if there's one really slow replica, then all the replica's timestamps could actually be off by `queue_len_response_deadline_s`.

Since `queue_len_response_deadline_s` is pretty low, this shouldn't be a major concern, but if it becomes larger, then the timestamps may be pretty inaccurate.

Reply:
yep you're right -- I have a follow-up item to generate all of these timestamps on the replica where possible